{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e9163091-b947-439e-b6a8-cea6c1ebd170",
   "metadata": {},
   "source": [
    "### This script digitizes sequences in the proteome and computes cross-correlations with a user-specified kernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2c65c61d-26b4-4b34-a22c-df0e61909c9e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\13kov\\anaconda3\\envs\\metapredict_env\\lib\\site-packages\\tqdm\\auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import matplotlib as mpl\n",
    "# mpl.use('Agg')\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42\n",
    "mpl.rcParams['text.usetex'] = False\n",
    "mpl.rcParams['font.sans-serif'] = 'Arial'\n",
    "mpl.rcParams['font.family'] = 'sans-serif'\n",
    "mpl.rcParams['figure.dpi'] = 300\n",
    "mpl.rcParams['image.interpolation'] = 'none'\n",
    "\n",
    "import os, re\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "from pprint import pprint\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "import logging\n",
    "logger = logging.Logger('catch_all')\n",
    "\n",
    "from Bio import SeqIO\n",
    "from scipy.interpolate import UnivariateSpline\n",
    "import matplotlib.transforms as mtransforms\n",
    "from matplotlib.patches import FancyBboxPatch\n",
    "\n",
    "import metapredict as meta\n",
    "# print(meta.__version__)\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "22e9b088-15d0-4e09-afb9-b8343627a26a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fasta_path = Path('uniprot-reviewed_yes+AND+proteome_up000005640.fasta')\n",
    "output_path = Path('/Users/13kov/Jupyter code/TF-RNA/ARM_search')\n",
    "\n",
    "if not output_path.is_dir():\n",
    "    output_path.mkdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d05e1f9b-3da5-4ffd-910a-954b529a54f3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def calc_tat_corr(seq): \n",
    "    # For correlation task, the TAT ARM sequence is: 49-RKKRRQRRR-57 \n",
    "\n",
    "    # Define kernel\n",
    "    # 1 represents a positive AA, 0  represents a neutral AA\n",
    "    ker=[1, 1, 1, 1, 1, 0, 1, 1, 1] # For Q=0\n",
    "#     ker=[1, 1, 1, 1, 1, 1, 1, 1, 1] # For Q=1\n",
    "    kmer = len(ker)\n",
    "    \n",
    "    # For filtering only on disordered regions\n",
    "    disorder_scores, disorder_regions = calc_disorder(seq)\n",
    "    disorder_threshold = 0.2 # Value of 0.2 suggested by Salman\n",
    "    \n",
    "    # Digitize the sequence based on charge to get a signal\n",
    "    sig=np.zeros(len(seq)).astype(int)\n",
    "    \n",
    "    for j, val in enumerate(seq):\n",
    "        if ((val=='R' or val=='K') and disorder_scores[j] >= disorder_threshold):\n",
    "            sig[j]=1\n",
    "        else:\n",
    "            sig[j]=0\n",
    "            \n",
    "    # Compute the cross-correlation\n",
    "    # Note that length of xcorr_vect will be 8 plus the length of the signal\n",
    "    xcorr_vect=scipy.signal.correlate(sig, ker, mode='full', method='direct')\n",
    "    \n",
    "    R_sig=np.zeros(len(seq)).astype(int)\n",
    "    \n",
    "    for j, val in enumerate(seq):\n",
    "        if ((val=='R') and disorder_scores[j] >= disorder_threshold):\n",
    "            R_sig[j]=1\n",
    "        else:\n",
    "            R_sig[j]=0\n",
    "            \n",
    "    ker_R=[1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
    "    xcorr_R=scipy.signal.correlate(R_sig, ker_R, mode='full', method='direct')   \n",
    "    xcorr_R[xcorr_R>=1]=1\n",
    "    \n",
    "    filt_vect=xcorr_vect*xcorr_R\n",
    "\n",
    "    max_corr=np.max(filt_vect)\n",
    "    sum_score=np.sum(filt_vect)\n",
    "                \n",
    "    return filt_vect, max_corr, sum_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3160f498-9d19-48c8-8c13-3f458dba9365",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_seqs(filt_vect,seq,gene_name,thresh):\n",
    "    # BCL11A - example of where blank sequences come up on C-terminus\n",
    "    # TFAP2A - example of where truncated sequences come up on N-terminus\n",
    "    out_dict={}\n",
    "    coord_dict={}\n",
    "    \n",
    "    for ind, val in enumerate(filt_vect):\n",
    "        \n",
    "        if val>=thresh and ind>=8:\n",
    "            \n",
    "            # Need to be very careful with the indexing here to get the right output!\n",
    "            sub_seq=seq[ind-8:ind+1] \n",
    "            \n",
    "            key=gene_name+'/'+str(ind)\n",
    "            \n",
    "            # Get rid of N-terminal tail sequences where the length of the ARM is not 9\n",
    "            if len(sub_seq)==9: \n",
    "                out_dict[key]=sub_seq+'/'+str(val)\n",
    "                \n",
    "                # Need to be careful here too\n",
    "                coord_dict[key]=ind-8\n",
    "\n",
    "    return out_dict, coord_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4f1c21e-9c58-4994-b215-51f8adfe7fab",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def calc_disorder(seq):\n",
    "    predictor = meta.predict_disorder_domains(seq)\n",
    "    \n",
    "    score = predictor.disorder\n",
    "    regions = predictor.disordered_domain_boundaries\n",
    "    \n",
    "    return score, regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "da67332d-90c8-4c45-94b4-675191a0c15f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 2 entries for DDIT3\n",
      "Found 2 entries for ZNF365\n",
      "Found 2 entries for CUX1\n",
      "TF max correlation file printed\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'gene_name'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m~\\anaconda3\\envs\\metapredict_env\\lib\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3360\u001b[0m             \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3361\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3362\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\metapredict_env\\lib\\site-packages\\pandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\metapredict_env\\lib\\site-packages\\pandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'gene_name'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp\\ipykernel_25404\\1594319597.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     59\u001b[0m \u001b[0mTF_seq_out\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'seq'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'x_corr'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTF_seq_out\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'seq'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'/'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexpand\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     60\u001b[0m \u001b[0mTF_seq_out\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'gene_name'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'trash'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTF_seq_out\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'gene_name'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'/'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexpand\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 61\u001b[1;33m \u001b[0mTF_seq_out\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'coord'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTF_coord_out\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'gene_name'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     62\u001b[0m \u001b[0mTF_seq_out\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'trash'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0minplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     63\u001b[0m \u001b[0mTF_seq_out\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mset_index\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'gene_name'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mdrop\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0minplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\metapredict_env\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3456\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3457\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3458\u001b[1;33m             \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3459\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3460\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\metapredict_env\\lib\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3361\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3362\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3363\u001b[1;33m                 \u001b[1;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3364\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3365\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0misna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhasnans\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'gene_name'"
     ]
    }
   ],
   "source": [
    "# Run TFs\n",
    "\n",
    "TF_df = pd.read_csv('TF_list.csv',header=None).T\n",
    "TF_query_list=TF_df.loc[0].tolist()\n",
    "# TF_query_list = ['BCL11A'] # Use this to debug specific sequences\n",
    "TF_full_df = pd.DataFrame()\n",
    "\n",
    "TF_out = {} # Declare dictionary\n",
    "TF_seq = {} # Declare dictionary\n",
    "TF_coord = {} # Declare dictionary\n",
    "\n",
    "for query in TF_query_list:\n",
    "#     print(query)\n",
    "    count = 0\n",
    "    with open(fasta_path, 'r') as file:\n",
    "        for record in SeqIO.parse(file, \"fasta\"):\n",
    "            description = record.description\n",
    "#             print(description)\n",
    "            gene_name = description[(description.find(\"GN=\")+3):description.find(\" PE\")]\n",
    "#             print(gene_name)\n",
    "            if query == gene_name:\n",
    "#                 print(f'Found {query} in {record.id}')  # This helps to see if there are multiple records for the same gene\n",
    "                ident = record.id\n",
    "                seq = str(record.seq).replace('U', 'C') # Replace U (selenocysteine) with C (cysteine)\n",
    "                count += 1\n",
    "                \n",
    "                # Make sure this if statement is one level down from the above\n",
    "                if count > 1:\n",
    "                    print(f'Found {count} entries for {query}')\n",
    "                elif count < 1:\n",
    "                #         raise Exception(f'Could not find entry for {query} in fasta file')\n",
    "                    print(f'Could not find entry for {query} in fasta file')\n",
    "                    \n",
    "                else:\n",
    "                    try:\n",
    "                        # TAT correlation\n",
    "                        xcorr_vect, max_corr, sum_score = calc_tat_corr(seq)\n",
    "                        TF_out[gene_name]=max_corr # Add max cross-correlation to dictionary\n",
    "\n",
    "                        # Print sequences\n",
    "                        thresh=5\n",
    "                        temp_seq,temp_coord=print_seqs(xcorr_vect,seq,gene_name,thresh)\n",
    "                        TF_seq.update(temp_seq) # Append temp dict to main\n",
    "                        TF_coord.update(temp_coord) # Append temp dict to main\n",
    "                        \n",
    "                    except BaseException as e:\n",
    "                        print(f'Error thrown by {gene_name}')\n",
    "                        logger.error(str(e))\n",
    "                                    \n",
    "# Print maximum xcorr\n",
    "TF_maxcorr_df=pd.DataFrame.from_dict(TF_out,orient='index',columns=['max_corr'])\n",
    "TF_maxcorr_df.to_csv('TF_maxcorr.csv')\n",
    "print('TF max correlation file printed')\n",
    "\n",
    "# Print full x-corr for sequences\n",
    "TF_seq_out=pd.DataFrame.from_dict(TF_seq,orient='index',columns=['seq'])\n",
    "TF_coord_out=pd.DataFrame.from_dict(TF_coord,orient='index',columns=['coord'])\n",
    "TF_seq_out['gene_name']=TF_seq_out.index\n",
    "TF_seq_out[['seq', 'x_corr']] = TF_seq_out['seq'].str.split('/', expand=True)\n",
    "TF_seq_out[['gene_name', 'trash']] = TF_seq_out['gene_name'].str.split('/', expand=True)\n",
    "TF_seq_out['coord_start']=TF_coord_out['coord']\n",
    "TF_seq_out['coord_end']=TF_coord_out['coord']+9\n",
    "TF_seq_out.drop('trash',axis=1,inplace=True)\n",
    "TF_seq_out.set_index('gene_name',drop=True,inplace=True)\n",
    "TF_seq_out.to_csv('TF_xcorr_strongarms.csv')\n",
    "print('TF correlation seq file printed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c9200ec8-214a-4cf1-8a09-5ab158ce614c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [16] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [12] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [21] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [16] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [4] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [16] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [20] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [20] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [16] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [15] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [16] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [2] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [17] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [16] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [16] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [15] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [5] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [17] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Warning: length of disorder [24] is <= window_size [24]. This happens when you have a small IDR relative to the minimum IDR size. Updating windowsize to match sequence length.\n",
      "Proteome file printed\n",
      "Proteome correlation seq file printed\n"
     ]
    }
   ],
   "source": [
    "# Run proteome\n",
    "\n",
    "proteome_output = {} # Declare dictionary for maxcorr\n",
    "proteome_seq={}\n",
    "\n",
    "with open(fasta_path, 'r') as file:\n",
    "    # Get proteome gene list from fasta file\n",
    "    for record in SeqIO.parse(file, \"fasta\"):\n",
    "        description = record.description\n",
    "        gene_name = description[(description.find(\"GN=\")+3):description.find(\" PE\")]\n",
    "        seq = str(record.seq).replace('U', 'C') # Replace U (selenocysteine) with C (cysteine)\n",
    "          \n",
    "    # To do: Include a column for the uniprot ID in the record.id  \n",
    "        \n",
    "        try:\n",
    "            # Tat correlation\n",
    "            xcorr_vect, max_corr, sum_score  = calc_tat_corr(seq)\n",
    "            proteome_output[gene_name] = max_corr\n",
    "            \n",
    "            # Print sequences\n",
    "            thresh=3\n",
    "            temp=print_seqs(xcorr_vect,seq,gene_name,thresh)\n",
    "            proteome_seq.update(temp) # Append temp dict to main\n",
    "            \n",
    "        except BaseException as e:\n",
    "            print(f'Error thrown by {gene_name}')\n",
    "            logger.error(str(e))\n",
    "\n",
    "# Print maximum xcorr\n",
    "proteome_intscore_df=pd.DataFrame.from_dict(proteome_output,orient='index',columns=['max_corr'])\n",
    "proteome_intscore_df.to_csv('proteome_maxcorr.csv')\n",
    "print('Proteome file printed')\n",
    "\n",
    "# Print full x-corr for sequences\n",
    "proteome_seq_out=pd.DataFrame.from_dict(proteome_seq,orient='index',columns=['seq'])\n",
    "proteome_seq_out['gene_name']=proteome_seq_out.index\n",
    "proteome_seq_out[['seq', 'x_corr']] = proteome_seq_out['seq'].str.split('/', expand=True)\n",
    "\n",
    "# Next 2 lines are a hack to get around the '/' character being located in some gene names\n",
    "proteome_seq_out['trash'] = proteome_seq_out['gene_name'].str.split('/', expand = True)[1]\n",
    "proteome_seq_out['gene_name'] = proteome_seq_out['gene_name'].str.split('/', expand = True)[0]\n",
    "\n",
    "proteome_seq_out.drop('trash',axis=1,inplace=True)\n",
    "proteome_seq_out.set_index('gene_name',drop=True,inplace=True)\n",
    "proteome_seq_out.to_csv('proteome_xcorr_allseq.csv')\n",
    "print('Proteome correlation seq file printed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5af48609-e79c-4695-b51b-15fb70378279",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "prot_df = pd.read_csv('proteome_maxcorr.csv',index_col=0)\n",
    "# RBP_df = pd.read_csv('RBP_list.csv',index_col=0)\n",
    "TF_df = pd.read_csv('TF_maxcorr.csv',index_col=0)\n",
    "\n",
    "# Filter TFs out of the whole proteome df\n",
    "prot_index = prot_df.index\n",
    "TF_index = TF_df.index\n",
    "mask1 = ~prot_index.isin(TF_index)\n",
    "mask2= prot_index.isin(TF_index)\n",
    "prot_filt = prot_df.loc[mask1]\n",
    "TF_only=prot_df.loc[mask2]\n",
    "\n",
    "# Filter TFs out of the RBP df\n",
    "# RBP_index = RBP_df.index\n",
    "# mask3 = prot_index.isin(RBP_index)\n",
    "# RBP_only = prot_df.loc[mask3]\n",
    "\n",
    "prot_filt.to_csv('proteome_maxcorr_TFfilt.csv')\n",
    "# TF_only.to_csv('TF_maxcorr.csv')\n",
    "# RBP_only.to_csv('RBP_intscore.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecad3dfb-c29a-4761-ba90-6b878d56d22a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "metapredict_env",
   "language": "python",
   "name": "metapredict_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
